### DISCUSSION TITLE
Allow deconstruction with `is {} (prop1, prop2...)` pattern

### DISCUSSION THREAD

**Discussion by virzak** [dotnet/csharplang#5004] [Language Ideas]
This could result in smaller, more readable code:

```cs
private void Proposal(MyClass? myNullableInstance)
{
    // OK
    if (myNullableInstance is { } myInstance)
    {
        var (name, age) = myInstance;
    }

    // Compiler Error: CS0103
    if (myNullableInstance is { } (name, age))
    {
    }
}
```

Where MyClass is a record
```cs
record MyClass(string Name, int Age);
```

or a class that implements a single Deconstruct method 
```cs
class MyClass
{
    public string Name { get; set; } = "";
    public int Age { get; set; }

    public void Deconstruct(out string name, out int age)
    {
        name = Name;
        age = Age;
    }
}
```

**Comment by 333fred** [comment]
You can just do

```cs
if (myNullableInstance is (name, age))
```

**Reply by virzak** [reply]
That doesn't compile.

**Reply by svick** [reply]
[This does:](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4GIB2BXANlghgIywgAII0CiBYAKBoAEAmYgYRoG8biviAHWMXMBJ0ALMQAKUAPY8pAZ1xYAFAFkEzPHLkB+YgFsEAOWx5CEAJJo5wXGgDGEAJSduHatw/F4xJQeM4KCysbexIYOWIBKB9yPQgAGmJcAHMnZ3dPLjdMrgBfFzyafNpqKAg7KSgAE2I1DVwtJToARgAGYkNcOMSYNGBiAEFUxwBuIA===)

```c#
if (myNullableInstance is var (name, age))
```

**Reply by virzak** [reply]
> 
> 
> [This does:](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4GIB2BXANlghgIywgAII0CiBYAKBoAEAmYgYRoG8biviAHWMXMBJ0ALMQAKUAPY8pAZ1xYAFAFkEzPHLkB+YgFsEAOWx5CEAJJo5wXGgDGEAJSduHatw/F4xJQeM4KCysbexIYOWIBKB9yPQgAGmJcAHMnZ3dPLjdMrgBfFzyafNpqKAg7KSgAE2I1DVwtJToARgAGYkNcOMSYNGBiAEFUxwBuIA===)
> 
> ```cs
> if (myNullableInstance is var (name, age))
> ```

Awesome!!! Didn't know about it, and I'll make use of it.

The original proposal is a lot less important now, but I wonder if it should still be implemented for consistency sake.

**Reply by HaloFour** [reply]
The `{ }` is just a degenerate form of the property pattern which can already be combined with positional patterns, it just has to come after it:

```
if (myNullableInstance is (var name, var age) { })
```

The positional pattern already performs a null check where appropriate, so you don't need to use a property pattern here.
